runtime/src/accounts_db.rs

```diff
@@ -2896,33 +2940,36 @@ impl AccountsDB {
                         let entry = accum
                             .entry(stored_account.meta.pubkey)
                             .or_insert_with(Vec::new);
+                        self.accounts_index.update_secondary_indexes(
+                            &stored_account.meta.pubkey,
+                            *slot,
+                            &stored_account.account_meta.owner,
+                            &stored_account.data,
+                            &self.account_indexes,
+                        );
                         entry.push((stored_account.meta.write_version, account_info));
                     },
                 );
 
             let mut accounts_map: HashMap<Pubkey, Vec<(u64, AccountInfo)>> = HashMap::new();
-            for accumulator_entry in accumulator.iter() {
+            for accumulator_entry in accumulator.into_iter() {
                 for (pubkey, storage_entry) in accumulator_entry {
-                    let entry = accounts_map.entry(*pubkey).or_insert_with(Vec::new);
-                    entry.extend(storage_entry.iter().cloned());
+                    let entry = accounts_map.entry(pubkey).or_insert_with(Vec::new);
+                    entry.extend(storage_entry.into_iter());
                 }
             }
 
             // Need to restore indexes even with older write versions which may
             // be shielding other accounts. When they are then purged, the
             // original non-shielded account value will be visible when the account
             // is restored from the append-vec
-            if !accumulator.is_empty() {
+            if !accounts_map.is_empty() {
                 let mut _reclaims: Vec<(u64, AccountInfo)> = vec![];
-                for (pubkey, account_infos) in accounts_map.iter_mut() {
+                for (pubkey, mut account_infos) in accounts_map.into_iter() {
                     account_infos.sort_by(|a, b| a.0.cmp(&b.0));
                     for (_, account_info) in account_infos {
-                        self.accounts_index.upsert(
-                            *slot,
-                            pubkey,
-                            account_info.clone(),
-                            &mut _reclaims,
-                        );
+                        self.accounts_index
+                            .upsert(*slot, &pubkey, account_info, &mut _reclaims);
```

Comment:
how about moving `update_secondary_indexes` https://github.com/solana-labs/solana/pull/14212/files#diff-1090394420d51617f3233275c2b65ed706b35b53b115fe65f82c682af8134a6fR2943 here just align with `AccountsDB::update_index`? Or is there any reason, we must do `update_secondary_indexes` earlier than this `upsert` for the snapshot (generate_index) case?

Moreover, I think we could even move the invocation of `update_secondary_indexes` inside `upsert` itself. So, this will clarify that we're updating primary index and secondary index at the same time.

Reply:
@ryoqun that's a good point, and one I was thinking about as well when initially implementing it. I think ideally the `update_secondary_indexes ` is called inside `upsert`, but my primary challenge came from the `generate_index` function  Note in that function, here is the update to the secondary index: https://github.com/solana-labs/solana/blob/f4ff936fb8ff64ebec752fc28c0cafde3879f16b/runtime/src/accounts_db.rs#L2943-L2949, and here is the update to the original index: https://github.com/solana-labs/solana/blob/f4ff936fb8ff64ebec752fc28c0cafde3879f16b/runtime/src/accounts_db.rs#L2972

Concerns:
1) `update_secondary_indexes` requires the account data, which is only available as a reference in the scan here, https://github.com/solana-labs/solana/blob/f4ff936fb8ff64ebec752fc28c0cafde3879f16b/runtime/src/accounts_db.rs#L2943-L2949.
So we would either have to clone out this account data so that the later call to `upsert` can be passed that account data, or:
2) We move upsert inside the scan there. Functionally this should be equivalent, as long as `upsert` is thread safe for the `par_iter` inside `scan_account_storage_inner`. I wasn't 100% sure this was the case, so I decided to mitigate risk and not do that here.

Reply:
Edit: I think calling the `update_secondary_indexes` where I have it right now is indeed not safe because it doesn't respect the `stored_account.meta.write_version`, i.e. this sort https://github.com/solana-labs/solana/blob/f4ff936fb8ff64ebec752fc28c0cafde3879f16b/runtime/src/accounts_db.rs#L2969 is critical in knowing how to order the updates to the index. Ordering is important because later updates will pop off earlier ones in the same slot.

Reply:
Updated here: https://github.com/solana-labs/solana/pull/14212/commits/84d82f5ecd3da227f9963275a159d44a37960e17